src/grpc/rpc_server.cpp

```diff
@@ -155,11 +155,23 @@ bool GrpcServer::run_generic_handler_cb(const std::string& rpc_name, boost::intr
             // respond immediately
             return true;
         }
-        cb = it->second;
+        cb = it->second.first;
     }
     return cb(rpc_data);
 }
 
+void GrpcServer::run_generic_completion_cb(const std::string& rpc_name,
+                                           boost::intrusive_ptr< GenericRpcData >& rpc_data) {
+    generic_rpc_completed_cb_t cb;
+    {
+        std::shared_lock< std::shared_mutex > lock(m_generic_rpc_registry_mtx);
+        auto it = m_generic_rpc_registry.find(rpc_name);
+        LOGMSG_ASSERT(it != m_generic_rpc_registry.end(), "completion cb not found for rpc {}", rpc_name);
+        cb = it->second.second;
+    }
+    if (cb) { cb(rpc_data); }
```

Comment:
We actually can avoid this additional lock and lookup, because completion can never be before the handling. In that case, handler can attach the cb reference in the rpc_data and then use that to do caller, right? Am I missing something?